src/tasks/LibraryBuilder/LibraryBuilder.cs

```diff
@@ -226,6 +253,79 @@ private static void WriteLinkerScriptFile(string exportsFile, List<string> expor
+        string runtimeConfig = GenerateRuntimeConfig();
+        File.WriteAllText(Path.Combine(OutputDirectory, "autoinit.c"),
+            Utils.GetEmbeddedResource("autoinit.c")
+                .Replace("%APPCTX_ENV_VARIABLES%", appContextEnvVariables)
+                .Replace("%ASSEMBLIES_LOADER%", assembliesLoader)
+                .Replace("%ASSETS_PATH%", assetsPath)
+                .Replace("%RUNTIME_CONFIG%", runtimeConfig));
+    }
+
+    private string GenerateAssembliesLoader()
+    {
+        var assembliesLoader = new StringBuilder();
+        foreach (string exportedAssembly in exportedAssemblies)
+        {
+            assembliesLoader.AppendLine($"    mono_assembly_open(\"{exportedAssembly}\", NULL);");
+        }
+        return assembliesLoader.ToString();
+    }
+
+    private string GenerateAssetsPath()
+    {
+        return AssetsPath ?? "DOTNET_ASSETS_PATH";
+    }
+
+    private string GenerateRuntimeConfig()
+    {
+        if (string.IsNullOrEmpty(RuntimeConfigBinFile))
+            return "    return;";
+
+        var runtimeConfig = new StringBuilder();
+        runtimeConfig.Append($"    char *file_name = {RuntimeConfigBinFile};");
+        runtimeConfig.Append("    int str_len = strlen (bundle_path) + strlen (file_name) + 1; // +1 is for the \"/\"");
+        runtimeConfig.Append("    char *file_path = (char *)malloc (sizeof (char) * (str_len +1)); // +1 is for the terminating null character");
+        runtimeConfig.Append("    int num_char = snprintf (file_path, (str_len + 1), \"%s/%s\", bundle_path, file_name);");
+        runtimeConfig.Append("    struct stat buffer;\n");
+        runtimeConfig.Append("    assert (num_char > 0 && num_char == str_len);\n");
+        runtimeConfig.Append("    if (stat (file_path, &buffer) == 0) {");
+        runtimeConfig.Append("        MonovmRuntimeConfigArguments *arg = (MonovmRuntimeConfigArguments *)malloc (sizeof (MonovmRuntimeConfigArguments));");
+        runtimeConfig.Append("        arg->kind = 0;");
+        runtimeConfig.Append("        arg->runtimeconfig.name.path = file_path;");
+        runtimeConfig.Append("        monovm_runtimeconfig_initialize (arg, cleanup_runtime_config, file_path);");
+        runtimeConfig.Append("    } else {");
+        runtimeConfig.Append("        free (file_path);");
+        runtimeConfig.Append("    }");
+        return runtimeConfig.ToString();
+    }
+
+    private string GenerateAppContextEnvVariables()
+    {
+        if (AppContextKeys.Length != AppContextValues.Length)
+        {
+            throw new LogAsErrorException($"'{nameof(AppContextKeys)}' length does not match '{nameof(AppContextValues)}' length. {AppContextKeys.Length} != {AppContextValues.Length}");
+        }
+
+        int numArgs = AppContextKeys.Length;
+        var appContextEnvVariables = new StringBuilder();
+
+        appContextEnvVariables.AppendLine($"    const char **appctx_keys, **appctx_values = (char**)malloc({numArgs} * sizeof(char*));");
```

Comment:
Same, why can't we do this in autoinit.c?

Reply:
I added this in LibraryBuilder because the amount of memory to allocate depends on the number of variables.

Or would it be preferred to have something like 
autoinit.c
```
    const char **appctx_keys, **appctx_values = (char**)malloc({%NUM_ARGS%} * sizeof(char*));
    ...
    monovm_initialize(%NUM_ARGS%, appctx_keys, appctx_values);
```

LibraryBuilder.cs
```
    private void WriteAutoInitializationFromTemplate()
    {
        string appContextEnvVariables = GenerateAppContextEnvVariables();
        string assembliesLoader = GenerateAssembliesLoader();
        string assetsPath = GenerateAssetsPath();
        string runtimeConfig = GenerateRuntimeConfig();
        File.WriteAllText(Path.Combine(OutputDirectory, "autoinit.c"),
            Utils.GetEmbeddedResource("autoinit.c")
                .Replace("%NUM_ARGS%", AppContextKeys.Length) // And checking AppContextKeys.Length != AppContextValues.Length much earlier
        ...
```

Reply:
@lateralusX does nativeaot allow extra app context values like this?

Reply:
Looking through the bootstrapping code, it doesn't look like its using the app context properties there, the bootstrapper pretty much calls into RhInitialize that calls InitDLL and that setup the runtime. Having that said it looks like they run __managed__Startup as part of library mode that does some additional initialization in managed code added by IL compiler:

compilationRoots.Add(new MainMethodRootProvider(entrypointModule, CreateInitializerList(typeSystemContext)));
compilationRoots.Add(new RuntimeConfigurationRootProvider(_runtimeOptions));
compilationRoots.Add(new ExpectedIsaFeaturesRootProvider(instructionSetSupport));

Looks like the initializer list also add app context settings into a init method that gets run at this stage and app context switches is arguments passed to IL compiler:

// If there are any AppContext switches the user wishes to enable, generate code that sets them.
if (_appContextSwitches.Count > 0)
{
    MethodDesc appContextInitMethod = new Internal.IL.Stubs.StartupCode.AppContextInitializerMethod(
    context.GeneratedAssembly.GetGlobalModuleType(), _appContextSwitches);
    initializerList.Add(appContextInitMethod);
}

and argument examples to IL compiler:

<IlcArg Include="@(RuntimeHostConfigurationOption->'--appcontextswitch:%(Identity)=%(Value)')" />
<IlcArg Include="--appcontextswitch:RUNTIME_IDENTIFIER=$(RuntimeIdentifier)" />

So you can set app context properties through IL compiler arguments. I guess we can do the same as msbuild arguments passed down to library builders from our targets.

Reply:
Changed to not allow any extra app context entries for now. We can add more if there is a need to install more environment variables that aren't passable into the generated runtimeconfig.json.